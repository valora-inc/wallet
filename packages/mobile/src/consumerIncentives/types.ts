export enum SuperchargeToken {
  cUSD = 'cUSD',
  cEUR = 'cEUR',
  cREAL = 'cREAL',
}

export interface SuperchargeTokenConfig {
  token: SuperchargeToken
  minBalance: number
  maxBalance: number
}

export interface SuperchargePendingReward {
  amount: string
  contractAddress: string
  createdAt: number
  index: number
  proof: string[]
  tokenAddress: string
}
