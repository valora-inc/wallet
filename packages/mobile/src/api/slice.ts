import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { SuperchargePendingReward } from 'src/consumerIncentives/types'
import config from 'src/geth/networkConfig'
import useSelector from 'src/redux/useSelector'
import { walletAddressSelector } from 'src/web3/selectors'

export const cloudFunctionsApi = createApi({
  reducerPath: 'cloudFunctionsApi',
  baseQuery: fetchBaseQuery({ baseUrl: config.cloudFunctionsUrl }),
  tagTypes: ['Supercharge'],
  endpoints: (builder) => ({
    fetchSuperchargeRewards: builder.query<SuperchargePendingReward[], string>({
      query: (address) => `fetchAvailableSuperchargeRewards?address=${address}`,
      transformResponse: (response: { availableRewards: SuperchargePendingReward[] }) =>
        response?.availableRewards,
      providesTags: ['Supercharge'],
    }),
  }),
})

// @ts-ignore This works but throws a build error because we're using an old typescript version :(
const { useFetchSuperchargeRewardsQuery } = cloudFunctionsApi

export function useFetchSuperchargeRewards() {
  const address = useSelector(walletAddressSelector)
  const { data, isLoading, isError } = useFetchSuperchargeRewardsQuery(address ?? '', {
    refetchOnMountOrArgChange: true,
  })
  return { superchargeRewards: data ?? [], isLoading, isError }
}
