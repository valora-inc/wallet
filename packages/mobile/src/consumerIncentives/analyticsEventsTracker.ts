import { RewardsEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { Screens } from 'src/navigator/Screens'

export enum RewardsScreenOrigin {
  RewardPill = 'RewardPill',
  NotificationBox = 'NotificationBox',
  RewardAvailableNotification = 'RewardAvailableNotification',
  PaymentDetail = 'PaymentDetail',
  PushNotification = 'PushNotification',
  SideMenu = 'SideMenu',
}

export enum RewardsScreenCta {
  ClaimRewards = 'ClaimRewards',
  CashIn = 'CashIn',
  VerifyPhone = 'VerifyPhone',
}

export const trackRewardsScreenOpenEvent = (uri: string, origin: RewardsScreenOrigin) => {
  if (uri.includes(Screens.ConsumerIncentivesHomeScreen)) {
    ValoraAnalytics.track(RewardsEvents.rewards_screen_opened, {
      origin,
    })
  }
}
