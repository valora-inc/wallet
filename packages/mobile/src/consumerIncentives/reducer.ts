import { createAction, createReducer } from '@reduxjs/toolkit'
import { SuperchargePendingReward } from 'src/consumerIncentives/types'

export interface State {
  loading: boolean
  error: boolean
}

export const initialState = {
  loading: false,
  error: false,
}

export const claimRewards = createAction<SuperchargePendingReward[]>('SUPERCHARGE/CLAIM_REWARDS')
export const claimRewardsSuccess = createAction('SUPERCHARGE/CLAIM_REWARDS_SUCCESS')
export const claimRewardsFailure = createAction('SUPERCHARGE/CLAIM_REWARDS_FAILURE')

export const reducer = createReducer(initialState, (builder) => {
  builder
    .addCase(claimRewards, (state) => ({
      ...state,
      loading: true,
      error: false,
    }))
    .addCase(claimRewardsSuccess, (state) => ({
      ...state,
      loading: false,
      error: false,
    }))
    .addCase(claimRewardsFailure, (state) => ({
      ...state,
      loading: false,
      error: true,
    }))
})
