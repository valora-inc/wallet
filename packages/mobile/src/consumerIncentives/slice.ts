import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { SuperchargePendingReward } from 'src/consumerIncentives/types'

export interface State {
  loading: boolean
  error: boolean
  availableRewards: SuperchargePendingReward[]
  fetchAvailableRewardsLoading: boolean
  fetchAvailableRewardsError: boolean
}

export const initialState: State = {
  loading: false,
  error: false,
  fetchAvailableRewardsLoading: false,
  fetchAvailableRewardsError: false,
  availableRewards: [],
}

const slice = createSlice({
  name: 'supercharge',
  initialState,
  reducers: {
    claimRewards: (state, action: PayloadAction<SuperchargePendingReward[]>) => ({
      ...state,
      loading: true,
      error: false,
    }),
    claimRewardsSuccess: (state) => ({
      ...state,
      loading: false,
      error: false,
    }),
    claimRewardsFailure: (state) => ({
      ...state,
      loading: false,
      error: true,
    }),
    fetchAvailableRewards: (state) => ({
      ...state,
      fetchAvailableRewardsLoading: true,
      fetchAvailableRewardsError: false,
    }),
    fetchAvailableRewardsSuccess: (state) => ({
      ...state,
      fetchAvailableRewardsLoading: false,
      fetchAvailableRewardsError: false,
    }),
    fetchAvailableRewardsFailure: (state) => ({
      ...state,
      fetchAvailableRewardsLoading: false,
      fetchAvailableRewardsError: true,
    }),
    setAvailableRewards: (state, action: PayloadAction<SuperchargePendingReward[]>) => ({
      ...state,
      availableRewards: action.payload,
    }),
  },
})

export const {
  claimRewards,
  claimRewardsSuccess,
  claimRewardsFailure,
  fetchAvailableRewards,
  fetchAvailableRewardsSuccess,
  fetchAvailableRewardsFailure,
  setAvailableRewards,
} = slice.actions

export default slice.reducer
