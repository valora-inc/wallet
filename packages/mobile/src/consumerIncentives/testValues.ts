import 'react-native'
import { SuperchargePendingReward } from 'src/consumerIncentives/types'
import { mockCusdAddress } from 'test/values'

export const ONE_CUSD_REWARD_RESPONSE: SuperchargePendingReward[] = [
  {
    contractAddress: '0xdistributorContract',
    tokenAddress: mockCusdAddress,
    amount: (1e18).toString(16),
    index: 0,
    proof: [],
    createdAt: 1645591363099,
  },
]
