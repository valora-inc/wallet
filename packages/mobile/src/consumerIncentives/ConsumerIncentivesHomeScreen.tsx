import Button, { BtnSizes } from '@celo/react-components/components/Button'
import Touchable from '@celo/react-components/components/Touchable'
import Times from '@celo/react-components/icons/Times'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import { StackScreenProps } from '@react-navigation/stack'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { Image, ScrollView, StyleSheet, Text, View } from 'react-native'
import { TouchableOpacity } from 'react-native-gesture-handler'
import { SafeAreaView, useSafeAreaInsets } from 'react-native-safe-area-context'
import { RewardsEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { CELO_REWARDS_T_AND_C } from 'src/brandingConfig'
import { RewardsScreenCta } from 'src/consumerIncentives/analyticsEventsTracker'
import { Namespaces } from 'src/i18n'
import { earn1, earn2, earn3, earnMain } from 'src/images/Images'
import { noHeader } from 'src/navigator/Headers'
import { navigate, navigateBack } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'
import useSelector from 'src/redux/useSelector'

type Props = StackScreenProps<StackParamList, Screens.ConsumerIncentivesHomeScreen>
export default function ConsumerIncentivesHomeScreen(props: Props) {
  const { t } = useTranslation(Namespaces.consumerIncentives)
  const userIsVerified = useSelector((state) => state.app.numberVerified)
  const insets = useSafeAreaInsets()

  const { rewardsPercent, rewardsMax } = useSelector((state) => state.app)

  const onPressCTA = () => {
    if (userIsVerified) {
      navigate(Screens.FiatExchangeOptions, { isCashIn: true })
    } else {
      navigate(Screens.VerificationEducationScreen, { hideOnboardingStep: true })
    }
    ValoraAnalytics.track(RewardsEvents.rewards_screen_cta_pressed, {
      buttonPressed: userIsVerified ? RewardsScreenCta.CashIn : RewardsScreenCta.VerifyPhone,
    })
  }

  const onLearnMore = () => navigate(Screens.WebViewScreen, { uri: CELO_REWARDS_T_AND_C })

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView
        contentContainerStyle={styles.contentContainer}
        showsVerticalScrollIndicator={false}
      >
        <Touchable
          style={[styles.closeButton, { marginTop: insets.top }]}
          onPress={navigateBack}
          borderless={true}
          hitSlop={variables.iconHitslop}
        >
          <Times />
        </Touchable>
        <Image source={earnMain} />
        <Text style={styles.title}>{t('title')}</Text>
        <Text style={styles.description}>{t('summary', { percent: rewardsPercent })}</Text>
        <View style={styles.section}>
          <Image source={earn1} style={styles.sectionImage} resizeMode="contain" />
          <View style={styles.sectionText}>
            <Text style={fontStyles.regular600}>{t('earnWeekly.header')}</Text>
            <Text style={fontStyles.small}>{t('earnWeekly.text')}</Text>
          </View>
        </View>
        <View style={styles.section}>
          <Image source={earn2} style={styles.sectionImage} resizeMode="contain" />
          <View style={styles.sectionText}>
            <Text style={fontStyles.regular600}>{t('noMinCommitment.header')}</Text>
            {userIsVerified ? (
              <Text style={fontStyles.small}>{t('noMinCommitment.earningText')}</Text>
            ) : (
              <Text style={fontStyles.small}>{t('noMinCommitment.connectText')}</Text>
            )}
          </View>
        </View>
        <View style={styles.section}>
          <Image source={earn3} style={styles.sectionImage} resizeMode="contain" />
          <View style={styles.sectionText}>
            <Text style={fontStyles.regular600}>{t('saveMoreEarnMore.header')}</Text>
            <Text style={fontStyles.small}>
              {t('saveMoreEarnMore.text', { rewardsMax, percent: rewardsPercent })}
            </Text>
          </View>
        </View>
        <Text style={[styles.description, { marginTop: 24 }]}>{t('conclusion')}</Text>
        <TouchableOpacity onPress={onLearnMore} testID="ConsumerIncentives/learnMore">
          <Text style={styles.learnMore}>{t('global:learnMore')}</Text>
        </TouchableOpacity>
      </ScrollView>
      <View style={styles.buttonContainer}>
        <Button
          size={BtnSizes.FULL}
          text={userIsVerified ? t('addCusd') : t('connectNumber')}
          onPress={onPressCTA}
          testID="ConsumerIncentives/CTA"
        />
      </View>
    </SafeAreaView>
  )
}

ConsumerIncentivesHomeScreen.navOptions = noHeader

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
  },
  contentContainer: {
    alignItems: 'center',
    marginHorizontal: 24,
  },
  closeButton: {
    alignSelf: 'flex-start',
  },
  title: {
    ...fontStyles.h2,
    marginTop: 32,
    textAlign: 'center',
  },
  description: {
    ...fontStyles.regular,
    textAlign: 'center',
    marginTop: 12,
  },
  section: {
    flexDirection: 'row',
    marginTop: 24,
    flex: 1,
    width: '100%',
  },
  sectionImage: {
    marginRight: 12,
  },
  sectionText: {
    flexDirection: 'column',
    marginLeft: 12,
    flex: 1,
  },
  buttonContainer: {
    paddingVertical: 16,
    paddingHorizontal: 24,
    borderTopColor: colors.gray2,
    borderTopWidth: 1,
  },
  learnMore: {
    ...fontStyles.notificationHeadline,
    fontSize: 17,
    marginVertical: 24,
    color: colors.greenUI,
  },
})
