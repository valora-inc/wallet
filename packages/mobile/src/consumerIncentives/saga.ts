import { toTransactionObject } from '@celo/connect'
import { ContractKit } from '@celo/contractkit'
import BigNumber from 'bignumber.js'
import { all, call, put, select, spawn, takeEvery } from 'redux-saga/effects'
import merkleDistributor from 'src/abis/MerkleDistributor.json'
import { showError, showMessage } from 'src/alert/actions'
import { RewardsEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { cloudFunctionsApi } from 'src/api/slice'
import { ErrorMessages } from 'src/app/ErrorMessages'
import {
  claimRewards,
  claimRewardsFailure,
  claimRewardsSuccess,
} from 'src/consumerIncentives/slice'
import { WEI_PER_TOKEN } from 'src/geth/consts'
import i18n from 'src/i18n'
import { navigateHome } from 'src/navigator/NavigationService'
import { TokenBalances } from 'src/tokens/reducer'
import { tokensByAddressSelector } from 'src/tokens/selectors'
import { addStandbyTransaction } from 'src/transactions/actions'
import {
  newTransactionContext,
  TokenTransactionTypeV2,
  TransactionStatus,
} from 'src/transactions/types'
import Logger from 'src/utils/Logger'
import { getContractKit } from 'src/web3/contracts'
import { getConnectedUnlockedAccount } from 'src/web3/saga'
import { getContract } from 'src/web3/utils'

const TAG = 'SuperchargeRewardsClaimer'

export function* claimRewardsSaga({ payload: rewards }: ReturnType<typeof claimRewards>) {
  try {
    const kit: ContractKit = yield call(getContractKit)
    const walletAddress: string = yield call(getConnectedUnlockedAccount)
    const tokens: TokenBalances = yield select(tokensByAddressSelector)
    const baseNonce: number = yield call(
      // @ts-ignore I can't figure out the syntax for this, it works but TS complains :'(
      [kit.web3.eth, kit.web3.eth.getTransactionCount],
      walletAddress
    )

    Logger.debug(TAG, `Starting to claim ${rewards.length} rewards with baseNonce: ${baseNonce}`)

    const receivedRewards: {
      fundsSource: string
      amount: string
      tokenAddress: string
      txHash: string
    }[] = yield all(
      rewards.map(async (reward, index) => {
        Logger.debug(TAG, `Start claiming reward at index ${index}: ${JSON.stringify(reward)}`)
        const merkleContract = await getContract(merkleDistributor.abi, reward.contractAddress)
        const fundsSource = await merkleContract.methods.fundsSource().call()

        const claimTx = toTransactionObject(
          kit.connection,
          merkleContract.methods.claim(reward.index, walletAddress, reward.amount, reward.proof)
        )
        const receipt = await claimTx.sendAndWaitForReceipt({
          from: walletAddress,
          nonce: baseNonce + index,
        })
        Logger.info(TAG, `Claimed reward at index ${index}: ${JSON.stringify(receipt)}`)
        const amount = new BigNumber(reward.amount, 16).div(WEI_PER_TOKEN).toString()
        const tokenAddress = reward.tokenAddress.toLowerCase()
        ValoraAnalytics.track(RewardsEvents.claimed_reward, {
          amount,
          token: tokens[tokenAddress]?.symbol ?? '',
        })
        return {
          fundsSource: fundsSource.toLowerCase(),
          tokenAddress,
          amount,
          txHash: receipt.transactionHash,
        }
      })
    )
    for (const reward of receivedRewards) {
      yield put(
        addStandbyTransaction({
          context: newTransactionContext('Claim Reward', reward.txHash),
          type: TokenTransactionTypeV2.Received,
          status: TransactionStatus.Complete,
          value: reward.amount,
          tokenAddress: reward.tokenAddress,
          comment: '',
          timestamp: Math.floor(Date.now() / 1000),
          address: reward.fundsSource,
          hash: reward.txHash,
        })
      )
    }
    yield put(cloudFunctionsApi.util.invalidateTags(['Supercharge']))
    yield put(claimRewardsSuccess())
    yield put(showMessage(i18n.t('superchargeClaimSuccess')))
    navigateHome()
  } catch (error) {
    yield put(claimRewardsFailure())
    yield put(showError(ErrorMessages.SUPERCHARGE_CLAIM_FAILED))
    Logger.error(TAG, 'Error claiming rewards', error)
  }
}

export function* watchClaimRewards() {
  yield takeEvery(claimRewards.type, claimRewardsSaga)
}

export function* superchargeSaga() {
  yield spawn(watchClaimRewards)
}
