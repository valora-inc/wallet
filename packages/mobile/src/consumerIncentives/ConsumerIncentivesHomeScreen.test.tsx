import { fireEvent, render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { ReactTestInstance } from 'react-test-renderer'
import { MockStoreEnhanced } from 'redux-mock-store'
import { useFetchSuperchargeRewards } from 'src/api/slice'
import { SUPERCHARGE_T_AND_C } from 'src/config'
import ConsumerIncentivesHomeScreen from 'src/consumerIncentives/ConsumerIncentivesHomeScreen'
import { SuperchargePendingReward, SuperchargeToken } from 'src/consumerIncentives/types'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'
import { StoredTokenBalance } from 'src/tokens/reducer'
import { createMockStore } from 'test/utils'
import { mockCusdAddress } from 'test/values'

jest.mock('src/api/slice', () => ({
  ...(jest.requireActual('src/api/slice') as any),
  useFetchSuperchargeRewards: jest.fn(),
}))

interface TokenBalances {
  [address: string]: Partial<StoredTokenBalance> | undefined
}

const CUSD_BALANCE: TokenBalances = {
  [mockCusdAddress]: {
    balance: '50',
    symbol: SuperchargeToken.cUSD,
    isCoreToken: true,
  },
}
const NO_BALANCES: TokenBalances = {
  [mockCusdAddress]: {
    balance: '5',
    symbol: SuperchargeToken.cUSD,
    isCoreToken: true,
  },
}

const EMPTY_REWARDS_RESPONSE: SuperchargePendingReward[] = []
export const ONE_CUSD_REWARD_RESPONSE: SuperchargePendingReward[] = [
  {
    contractAddress: '0xdistributorContract',
    tokenAddress: mockCusdAddress,
    amount: (1e18).toString(16),
    index: 0,
    proof: [],
    createdAt: 1645591363099,
  },
]

function expectVisibleMainComponents(
  queryByTestId: (testId: string) => ReactTestInstance | null,
  mainComponentTestId: string
) {
  expect(queryByTestId('ConsumerIncentives/CTA')).toBeTruthy()
  expect(queryByTestId('LearnMore')).toBeTruthy()
  const componentTestIds = [
    'SuperchargeLoading',
    'ClaimSuperchargeDescription',
    'SuperchargingInfo',
    'SuperchargeInstructions',
  ]
  for (const testId of componentTestIds) {
    if (testId === mainComponentTestId) {
      expect(queryByTestId(testId)).not.toBeNull()
    } else {
      expect(queryByTestId(testId)).toBeNull()
    }
  }
}

describe('ConsumerIncentivesHomeScreen', () => {
  let store: MockStoreEnhanced<RootState, {}>
  beforeEach(() => mockQueryResponse(EMPTY_REWARDS_RESPONSE))

  function createStore({
    numberVerified,
    tokenBalances,
  }: {
    numberVerified: boolean
    tokenBalances: TokenBalances
  }) {
    store = createMockStore({
      app: {
        numberVerified,
        superchargeTokens: [
          {
            token: SuperchargeToken.cUSD,
            minBalance: 10,
            maxBalance: 1000,
          },
          {
            token: SuperchargeToken.cREAL,
            minBalance: 50,
            maxBalance: 6000,
          },
        ],
      },
      tokens: { tokenBalances },
    })
    return store
  }

  function mockQueryResponse(data: SuperchargePendingReward[]) {
    ;(useFetchSuperchargeRewards as jest.Mock).mockImplementation(() => ({
      superchargeRewards: data,
      isLoading: false,
      isError: false,
    }))
  }

  it('renders Supercharge instructions when not Supercharging', async () => {
    const { queryByTestId } = render(
      <Provider
        store={createStore({
          numberVerified: false,
          tokenBalances: NO_BALANCES,
        })}
      >
        <ConsumerIncentivesHomeScreen />
      </Provider>
    )

    expectVisibleMainComponents(queryByTestId, 'SuperchargeInstructions')
  })

  it('renders Supercharge Info when Supercharging', async () => {
    const { queryByTestId } = render(
      <Provider
        store={createStore({
          numberVerified: true,
          tokenBalances: CUSD_BALANCE,
        })}
      >
        <ConsumerIncentivesHomeScreen />
      </Provider>
    )

    expectVisibleMainComponents(queryByTestId, 'SuperchargingInfo')
  })

  it('renders available rewards to claim when they are available', async () => {
    mockQueryResponse(ONE_CUSD_REWARD_RESPONSE)
    const { queryByTestId } = render(
      <Provider
        store={createStore({
          numberVerified: true,
          tokenBalances: CUSD_BALANCE,
        })}
      >
        <ConsumerIncentivesHomeScreen />
      </Provider>
    )

    expectVisibleMainComponents(queryByTestId, 'ClaimSuperchargeDescription')
  })

  it('navigates to cash in screen if user is verified and CTA is tapped', async () => {
    const { getByTestId } = render(
      <Provider
        store={createStore({
          numberVerified: true,
          tokenBalances: NO_BALANCES,
        })}
      >
        <ConsumerIncentivesHomeScreen />
      </Provider>
    )

    fireEvent.press(getByTestId('ConsumerIncentives/CTA'))
    expect(navigate).toHaveBeenCalledWith(Screens.FiatExchangeOptions, { isCashIn: true })
  })

  it('navigates to Phone Confirmation screen if user is not verified and CTA is tapped', async () => {
    const { getByTestId } = render(
      <Provider
        store={createStore({
          numberVerified: false,
          tokenBalances: NO_BALANCES,
        })}
      >
        <ConsumerIncentivesHomeScreen />
      </Provider>
    )

    fireEvent.press(getByTestId('ConsumerIncentives/CTA'))

    expect(navigate).toHaveBeenCalledWith(Screens.VerificationEducationScreen, {
      hideOnboardingStep: true,
    })
  })

  it('opens a WebView when Learn More is tapped', async () => {
    const { getByTestId } = render(
      <Provider
        store={createStore({
          numberVerified: true,
          tokenBalances: NO_BALANCES,
        })}
      >
        <ConsumerIncentivesHomeScreen />
      </Provider>
    )

    fireEvent.press(getByTestId('LearnMore'))

    expect(navigate).toHaveBeenCalledWith(Screens.WebViewScreen, { uri: SUPERCHARGE_T_AND_C })
  })

  it('dispatches a claim rewards action when CTA is tapped', async () => {
    mockQueryResponse(ONE_CUSD_REWARD_RESPONSE)
    const { getByTestId } = render(
      <Provider
        store={createStore({
          numberVerified: true,
          tokenBalances: CUSD_BALANCE,
        })}
      >
        <ConsumerIncentivesHomeScreen />
      </Provider>
    )

    fireEvent.press(getByTestId('ConsumerIncentives/CTA'))

    expect(store.getActions()).toMatchInlineSnapshot(`
      Array [
        Object {
          "payload": Array [
            Object {
              "amount": "de0b6b3a7640000",
              "contractAddress": "0xdistributorContract",
              "createdAt": 1645591363099,
              "index": 0,
              "proof": Array [],
              "tokenAddress": "0x874069Fa1Eb16D44d622F2e0Ca25eeA172369bC1",
            },
          ],
          "type": "supercharge/claimRewards",
        },
      ]
    `)
  })
})
