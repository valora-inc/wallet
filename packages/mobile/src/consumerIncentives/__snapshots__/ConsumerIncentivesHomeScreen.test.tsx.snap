// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsumerIncentivesHomeScreen renders correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "marginHorizontal": 24,
      }
    }
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignSelf": "flex-start",
            },
            Object {
              "marginTop": 0,
            },
          ]
        }
      >
        <svg
          height={16}
          style={Object {}}
          viewBox="0 0 16 16"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13.9999 2.00146L1.99994 14.0015"
            stroke="#2E3338"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            style={Object {}}
          />
          <path
            d="M1.99994 2.00146L13.9999 14.0015"
            stroke="#2E3338"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            style={Object {}}
          />
        </svg>
      </View>
      <Image
        source={
          Object {
            "testUri": "../../../packages/mobile/src/images/earnMain.png",
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Jost-Medium",
            "fontSize": 22,
            "lineHeight": 28,
            "marginTop": 32,
            "textAlign": "center",
          }
        }
      >
        title
      </Text>
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Inter-Regular",
            "fontSize": 16,
            "lineHeight": 22,
            "marginTop": 12,
            "textAlign": "center",
          }
        }
      >
        summary, {"percent":5}
      </Text>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginTop": 24,
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../packages/mobile/src/images/earn1.png",
            }
          }
          style={
            Object {
              "marginRight": 12,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "marginLeft": 12,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              }
            }
          >
            earnWeekly.header
          </Text>
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": 18,
              }
            }
          >
            earnWeekly.text
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginTop": 24,
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../packages/mobile/src/images/earn2.png",
            }
          }
          style={
            Object {
              "marginRight": 12,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "marginLeft": 12,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              }
            }
          >
            noMinCommitment.header
          </Text>
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": 18,
              }
            }
          >
            noMinCommitment.earningText
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginTop": 24,
            "width": "100%",
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            Object {
              "testUri": "../../../packages/mobile/src/images/earn3.png",
            }
          }
          style={
            Object {
              "marginRight": 12,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "column",
              "marginLeft": 12,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              }
            }
          >
            saveMoreEarnMore.header
          </Text>
          <Text
            style={
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": 18,
              }
            }
          >
            saveMoreEarnMore.text, {"rewardsMax":1000,"percent":5}
          </Text>
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
              "marginTop": 12,
              "textAlign": "center",
            },
            Object {
              "marginTop": 24,
            },
          ]
        }
      >
        conclusion
      </Text>
      <RNGestureHandlerButton
        collapsable={false}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        testID="ConsumerIncentives/learnMore"
      >
        <View
          accessible={true}
          style={
            Array [
              Object {},
              Object {
                "opacity": 1,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#1AB775",
                "fontFamily": "Jost-Medium",
                "fontSize": 17,
                "lineHeight": 22,
                "marginVertical": 24,
              }
            }
          >
            global:learnMore
          </Text>
        </View>
      </RNGestureHandlerButton>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "borderTopColor": "#EDEEEF",
        "borderTopWidth": 1,
        "paddingHorizontal": 24,
        "paddingVertical": 16,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "column",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 100,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          nativeBackgroundAndroid={
            Object {
              "attribute": "selectableItemBackground",
              "rippleRadius": undefined,
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#1AB775",
              "flexGrow": 1,
              "height": 48,
              "justifyContent": "center",
              "opacity": undefined,
              "paddingHorizontal": 24,
              "paddingVertical": 5,
            }
          }
          testID="ConsumerIncentives/CTA"
        >
          <Text
            maxFontSizeMultiplier={1}
            style={
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              }
            }
          >
            addCusd
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
